In [21]:
import pandas as pd
import re

In [22]:
file_path = r'C:\Users\CS\Documents\Caltech AIML\NLP_Speech_Recognition_SEngine\TwitterHate.csv'

In [23]:
twitter_df = pd.read_csv(file_path)
print(twitter_df.head())

   id  label                                              tweet
0   1      0   @user when a father is dysfunctional and is s...
1   2      0  @user @user thanks for #lyft credit i can't us...
2   3      0                                bihday your majesty
3   4      0  #model   i love u take with u all the time in ...
4   5      0             factsguide: society now    #motivation


In [24]:
tweet_list = twitter_df['tweet'].tolist()

print(tweet_list[:5])

[' @user when a father is dysfunctional and is so selfish he drags his kids into his dysfunction.   #run', "@user @user thanks for #lyft credit i can't use cause they don't offer wheelchair vans in pdx.    #disapointed #getthanked", '  bihday your majesty', '#model   i love u take with u all the time in urð\x9f\x93±!!! ð\x9f\x98\x99ð\x9f\x98\x8eð\x9f\x91\x84ð\x9f\x91\x85ð\x9f\x92¦ð\x9f\x92¦ð\x9f\x92¦  ', ' factsguide: society now    #motivation']


In [36]:
print(twitter_df.columns)

Index(['id', 'label', 'tweet'], dtype='object')


In [25]:
# normalize the casing
tweet_list = [tweet.lower() for tweet in tweet_list]

print(tweet_list[:5])

[' @user when a father is dysfunctional and is so selfish he drags his kids into his dysfunction.   #run', "@user @user thanks for #lyft credit i can't use cause they don't offer wheelchair vans in pdx.    #disapointed #getthanked", '  bihday your majesty', '#model   i love u take with u all the time in urð\x9f\x93±!!! ð\x9f\x98\x99ð\x9f\x98\x8eð\x9f\x91\x84ð\x9f\x91\x85ð\x9f\x92¦ð\x9f\x92¦ð\x9f\x92¦  ', ' factsguide: society now    #motivation']


In [26]:
# Using regular expressions, remove user handles. These begin with '@’.

tweet_list = [re.sub(r'@[^\s]+', '', tweet) for tweet in tweet_list]

print(tweet_list[:5])

['  when a father is dysfunctional and is so selfish he drags his kids into his dysfunction.   #run', "  thanks for #lyft credit i can't use cause they don't offer wheelchair vans in pdx.    #disapointed #getthanked", '  bihday your majesty', '#model   i love u take with u all the time in urð\x9f\x93±!!! ð\x9f\x98\x99ð\x9f\x98\x8eð\x9f\x91\x84ð\x9f\x91\x85ð\x9f\x92¦ð\x9f\x92¦ð\x9f\x92¦  ', ' factsguide: society now    #motivation']


In [27]:
# Using regular expressions, remove URLs

tweet_list = [re.sub(r'https?:\/\/.*[\r\n]*', '', tweet) for tweet in tweet_list]

print(tweet_list[:5])

['  when a father is dysfunctional and is so selfish he drags his kids into his dysfunction.   #run', "  thanks for #lyft credit i can't use cause they don't offer wheelchair vans in pdx.    #disapointed #getthanked", '  bihday your majesty', '#model   i love u take with u all the time in urð\x9f\x93±!!! ð\x9f\x98\x99ð\x9f\x98\x8eð\x9f\x91\x84ð\x9f\x91\x85ð\x9f\x92¦ð\x9f\x92¦ð\x9f\x92¦  ', ' factsguide: society now    #motivation']


In [28]:
# Using TweetTokenizer from NLTK, tokenize the tweets into individual terms

from nltk.tokenize import TweetTokenizer
tokenizer = TweetTokenizer()
tweets_tokenized = [tokenizer.tokenize(tweet) for tweet in tweet_list]

print(tweets_tokenized[:5])

[['when', 'a', 'father', 'is', 'dysfunctional', 'and', 'is', 'so', 'selfish', 'he', 'drags', 'his', 'kids', 'into', 'his', 'dysfunction', '.', '#run'], ['thanks', 'for', '#lyft', 'credit', 'i', "can't", 'use', 'cause', 'they', "don't", 'offer', 'wheelchair', 'vans', 'in', 'pdx', '.', '#disapointed', '#getthanked'], ['bihday', 'your', 'majesty'], ['#model', 'i', 'love', 'u', 'take', 'with', 'u', 'all', 'the', 'time', 'in', 'urð', '\x9f', '\x93', '±', '!', '!', '!', 'ð', '\x9f', '\x98', '\x99', 'ð', '\x9f', '\x98', '\x8e', 'ð', '\x9f', '\x91', '\x84', 'ð', '\x9f', '\x91', 'ð', '\x9f', '\x92', '¦', 'ð', '\x9f', '\x92', '¦', 'ð', '\x9f', '\x92', '¦'], ['factsguide', ':', 'society', 'now', '#motivation']]


In [29]:
# Remove stop words

from nltk.corpus import stopwords
stop_words = set(stopwords.words('english'))

tweets_tokenized = [[word for word in tweet if word not in stop_words] for tweet in tweets_tokenized]

print(tweets_tokenized[:5])

[['father', 'dysfunctional', 'selfish', 'drags', 'kids', 'dysfunction', '.', '#run'], ['thanks', '#lyft', 'credit', "can't", 'use', 'cause', 'offer', 'wheelchair', 'vans', 'pdx', '.', '#disapointed', '#getthanked'], ['bihday', 'majesty'], ['#model', 'love', 'u', 'take', 'u', 'time', 'urð', '\x9f', '\x93', '±', '!', '!', '!', 'ð', '\x9f', '\x98', '\x99', 'ð', '\x9f', '\x98', '\x8e', 'ð', '\x9f', '\x91', '\x84', 'ð', '\x9f', '\x91', 'ð', '\x9f', '\x92', '¦', 'ð', '\x9f', '\x92', '¦', 'ð', '\x9f', '\x92', '¦'], ['factsguide', ':', 'society', '#motivation']]


In [30]:
# Remove redundant terms like ‘amp’, ‘rt’,"#" etc.

tweets_tokenized = [[word for word in tweet if word not in ['amp', 'rt', "#"]] for tweet in tweets_tokenized]

print(tweets_tokenized[:5])

[['father', 'dysfunctional', 'selfish', 'drags', 'kids', 'dysfunction', '.', '#run'], ['thanks', '#lyft', 'credit', "can't", 'use', 'cause', 'offer', 'wheelchair', 'vans', 'pdx', '.', '#disapointed', '#getthanked'], ['bihday', 'majesty'], ['#model', 'love', 'u', 'take', 'u', 'time', 'urð', '\x9f', '\x93', '±', '!', '!', '!', 'ð', '\x9f', '\x98', '\x99', 'ð', '\x9f', '\x98', '\x8e', 'ð', '\x9f', '\x91', '\x84', 'ð', '\x9f', '\x91', 'ð', '\x9f', '\x92', '¦', 'ð', '\x9f', '\x92', '¦', 'ð', '\x9f', '\x92', '¦'], ['factsguide', ':', 'society', '#motivation']]


In [31]:
# Extra cleanup by removing terms with a length of 1

tweets_tokenized = [[word for word in tweet if len(word) > 1] for tweet in tweets_tokenized]

print(tweets_tokenized[:5])

[['father', 'dysfunctional', 'selfish', 'drags', 'kids', 'dysfunction', '#run'], ['thanks', '#lyft', 'credit', "can't", 'use', 'cause', 'offer', 'wheelchair', 'vans', 'pdx', '#disapointed', '#getthanked'], ['bihday', 'majesty'], ['#model', 'love', 'take', 'time', 'urð'], ['factsguide', 'society', '#motivation']]


In [33]:
# Get all the tokenized terms into 1 large list

from collections import Counter

# Flatten the list of lists into a single list of terms

all_terms = [term for tweet in tweets_tokenized for term in tweet]

term_frequencies = Counter(all_terms)

# Get the 10 most common terms
top_10_terms = term_frequencies.most_common(10)

print(top_10_terms)

[('...', 2810), ('day', 2022), ('happy', 1568), ('#love', 1542), ('love', 1206), ('time', 1088), ("i'm", 1017), ('like', 973), ('today', 941), ('new', 924)]


In [37]:
# Data formatting for predictive modeling
# Join the tokens back to form string ( vectorizers)
from sklearn.model_selection import train_test_split

tweets_joined = [' '.join(tweet) for tweet in tweets_tokenized]

Y = twitter_df['label']
X = tweets_joined

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [38]:
# We’ll use TF-IDF values for the terms as a feature to get into a vector space model

from sklearn.feature_extraction.text import TfidfVectorizer

# Instantiate TF-IDF Vectorizer with a maximum of 5000 terms in the vocabulary
tfidf_vectorizer = TfidfVectorizer(max_features=5000)

# Fit and transform the vectorizer on the train set
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)

# Transform the test set
X_test_tfidf = tfidf_vectorizer.transform(X_test)

In [40]:
# Model building: Ordinary Logistic Regression
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Assuming X_train_tfidf, Y_train, X_test_tfidf, and Y_test are already defined

# Instantiate Logistic Regression with default parameters
log_reg = LogisticRegression()

# Fit the model to the train data
log_reg.fit(X_train_tfidf, Y_train)

# Make predictions for the train and the test set
train_predictions = log_reg.predict(X_train_tfidf)
test_predictions = log_reg.predict(X_test_tfidf)

# You can also evaluate the model using metrics like accuracy
train_accuracy = accuracy_score(Y_train, train_predictions)
test_accuracy = accuracy_score(Y_test, test_predictions)

print(f"Train Accuracy: {train_accuracy}")
print(f"Test Accuracy: {test_accuracy}")



Train Accuracy: 0.9563142868317103
Test Accuracy: 0.9505709369623025


Train Accuracy (95.63%) and Test Accuracy (95.06%) are both high: This suggests that the model is performing well in correctly classifying the tweets. It has learned the patterns in the training data effectively and is also generalizing well to new, unseen data (test set).

Similarity in Train and Test Accuracies: The close values indicate that there is no significant overfitting. Overfitting would be suggested if the train accuracy was significantly higher than the test accuracy.

In [41]:
from sklearn.metrics import accuracy_score, recall_score, f1_score

# Assuming log_reg is your trained Logistic Regression model
# and X_train_tfidf and Y_train are your training features and labels

# Predictions on the train set
train_predictions = log_reg.predict(X_train_tfidf)

# Accuracy on the train set
train_accuracy = accuracy_score(Y_train, train_predictions)

# Recall on the train set
train_recall = recall_score(Y_train, train_predictions)

# F1 score on the train set
train_f1_score = f1_score(Y_train, train_predictions)

print(f"Train Accuracy: {train_accuracy}")
print(f"Train Recall: {train_recall}")
print(f"Train F1 Score: {train_f1_score}")

# For the recall: 
# - 'High' generally means close to 1.0
# - 'Decent' might be around 0.5 to 0.7
# - 'Low' would be closer to 0.0


Train Accuracy: 0.9563142868317103
Train Recall: 0.39529675251959684
Train F1 Score: 0.5583234480031632


#### Accuracy: High (95.63%) - Generally good classification.
#### Recall: Low (39.53%) - Misses many true positives.
#### F1 Score: Moderate (55.83%) - Needs improvement in identifying positive cases.

In [42]:
# Asjust class imbalance

from sklearn.linear_model import LogisticRegression

# Adjust class_weight to 'balanced' to handle class imbalance
log_reg = LogisticRegression(class_weight='balanced')

# Fit the model to your training data
log_reg.fit(X_train_tfidf, Y_train)

LogisticRegression(class_weight='balanced')

In [43]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, recall_score, f1_score

# Adjusting the model for class imbalance
log_reg_balanced = LogisticRegression(class_weight='balanced')

# Training the model on the train set
log_reg_balanced.fit(X_train_tfidf, Y_train)

# Making predictions on the train set
train_predictions_balanced = log_reg_balanced.predict(X_train_tfidf)

# Evaluating the model
accuracy_balanced = accuracy_score(Y_train, train_predictions_balanced)
recall_balanced = recall_score(Y_train, train_predictions_balanced)
f1_score_balanced = f1_score(Y_train, train_predictions_balanced)

print(f"Balanced Train Accuracy: {accuracy_balanced}")
print(f"Balanced Train Recall: {recall_balanced}")
print(f"Balanced Train F1 Score: {f1_score_balanced}")

Balanced Train Accuracy: 0.9468497008095741
Balanced Train Recall: 0.9680851063829787
Balanced Train F1 Score: 0.7178741955574008


Comparing the new results with the adjusted class weights to the old results:

- **Balanced Accuracy (94.68%) vs. Old Accuracy (95.63%):**
  - *Slight decrease* in overall accuracy.
  - Still *high*, indicating good general classification ability.

- **Balanced Recall (96.81%) vs. Old Recall (39.53%):**
  - *Significant increase* in recall.
  - Indicates a *substantial improvement* in identifying true positive cases (e.g., better at detecting hate speech).

- **Balanced F1 Score (71.79%) vs. Old F1 Score (55.83%):**
  - *Notable improvement* in F1 score.
  - Suggests a *better balance* between precision and recall, especially in correctly identifying positive cases.

**Insights:**
- Adjusting for class imbalance leads to a more *sensitive model*, especially in terms of identifying the minority class.
- There's a *trade-off* between recall and precision, as seen in the increased recall and slightly decreased accuracy.
- The improved F1 score indicates a more *balanced and effective model* for this specific task (e.g., hate speech detection).


In [ ]:
# Regularization and Hyperparameter tuning:

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV, StratifiedKFold

# Define the parameter grid to search for 'C' and 'penalty' parameters
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'penalty': ['l1', 'l2']
}

# Instantiate Logistic Regression with balanced class weight
log_reg = LogisticRegression(class_weight='balanced')

# Use StratifiedKFold for handling class imbalance in cross-validation
stratified_kfold = StratifiedKFold(n_splits=5)

# Set up GridSearchCV
grid_search = GridSearchCV(estimator=log_reg, param_grid=param_grid, cv=stratified_kfold, scoring='f1')

# Fit GridSearchCV to the training data (use X_train_tfidf and Y_train)
grid_search.fit(X_train_tfidf, Y_train)

# After fitting, you can get the best parameters and best score
best_params = grid_search.best_params_
best_score = grid_search.best_score_



In [48]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV, StratifiedKFold

# Define the parameter grid to search for 'C' and 'penalty' parameters
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'penalty': ['l1', 'l2']
}

# Instantiate Logistic Regression with balanced class weight
log_reg = LogisticRegression(class_weight='balanced')

# Use StratifiedKFold with 4 folds for handling class imbalance in cross-validation
stratified_kfold = StratifiedKFold(n_splits=4)

# Set up GridSearchCV with 'recall' as the metric for scoring
grid_search_recall = GridSearchCV(estimator=log_reg, param_grid=param_grid, cv=stratified_kfold, scoring='recall')

# Fit GridSearchCV to the training data (use X_train_tfidf and Y_train)
grid_search_recall.fit(X_train_tfidf, Y_train)

# After fitting, you can get the best parameters and best recall score
best_params_recall = grid_search_recall.best_params_
best_recall_score = grid_search_recall.best_score_


c:\Users\CS\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
c:\Users\CS\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

In [49]:
from sklearn.metrics import recall_score, f1_score

# Assuming grid_search_recall has already been fitted to the training data

# Retrieve the best parameters
best_params = grid_search_recall.best_params_

# Use the best estimator from the grid search
best_estimator = grid_search_recall.best_estimator_

# Make predictions on the test set
test_predictions = best_estimator.predict(X_test_tfidf)

# Calculate the recall on the test set
test_recall = recall_score(Y_test, test_predictions)

# Calculate the F1 score on the test set
test_f1_score = f1_score(Y_test, test_predictions)

print(f"Best Parameters: {best_params}")
print(f"Test Recall: {test_recall}")
print(f"Test F1 Score: {test_f1_score}")


Best Parameters: {'C': 1, 'penalty': 'l2'}
Test Recall: 0.8004385964912281
Test F1 Score: 0.5978705978705979
